pool/position_test.gno

```diff
@@ -0,0 +1,112 @@
+		{validAddr, 100, 200, false, "", "ZzF3ZXNrYzZ0eWc5anhndWpsdGEwNDdoNmx0YTA0N2g2bGRqbHVkdV9fMTAwX18yMDA="},                                                                         // tickLower < tickUpper
+	}
+
+	for _, tc := range tests {
+		if tc.shouldPanic {
+			uassert.PanicsWithMessage(t, tc.panicMsg, func() { positionGetKey(tc.owner, tc.tickLower, tc.tickUpper) })
+		} else {
+			key := positionGetKey(tc.owner, tc.tickLower, tc.tickUpper)
+			uassert.Equal(t, tc.expectedKey, key)
+		}
+	}
+}
+
+func TestPositionUpdateWithKey(t *testing.T) {
+	var dummyPool *Pool
+	var positionKey string
+
+	t.Run("set up initial data for this test function", func(t *testing.T) {
+		dummyPool = newPool(
+			"token0",
+			"token1",
+			100,
+			10,
+			common.TickMathGetSqrtRatioAtTick(0),
+		)
+
+		positionKey = positionGetKey(
+			testutils.TestAddress("dummyAddr"),
+			100,
+			200,
+		)
+	})
+
+	tests := []struct {
+		liquidity         *i256.Int
+		amount0           *u256.Uint
+		amount1           *u256.Uint
+		shouldPanic       bool
+		panicMsg          string
+		expectedLiquidity string
+	}{
+		{i256.MustFromDecimal("0"), u256.Zero(), u256.Zero(), true, `[GNOSWAP-POOL-010] zero liquidity || position.gno__positionUpdate() || both liquidityDelta and current position's liquidity are zero`, ""},
+		{i256.MustFromDecimal("100000"), u256.Zero(), u256.Zero(), false, "", "100000"},
+	}
+
+	for _, tc := range tests {
+		if tc.shouldPanic {
+			uassert.PanicsWithMessage(t, tc.panicMsg, func() { dummyPool.positionUpdateWithKey(positionKey, tc.liquidity, tc.amount0, tc.amount1) })
+		} else {
+			newPos := dummyPool.positionUpdateWithKey(positionKey, tc.liquidity, tc.amount0, tc.amount1)
+			uassert.Equal(t, newPos.liquidity.ToString(), tc.expectedLiquidity)
+		}
+	}
+}
+
+func TestPositionUpdate(t *testing.T) {
+	position := PositionInfo{
+		liquidity: u256.Zero(),
+	}
+
+	tests := []struct {
+		liquidityDelta       *i256.Int
+		feeGrowthInside0X128 *u256.Uint
+		feeGrowthInside1X128 *u256.Uint
+		shouldPanic          bool
+		panicMsg             string
+		expectedLiquidity    string
+	}{
+		{i256.MustFromDecimal("0"), u256.Zero(), u256.Zero(), true, `[GNOSWAP-POOL-010] zero liquidity || position.gno__positionUpdate() || both liquidityDelta and current position's liquidity are zero`, ""},
+		{i256.MustFromDecimal("100000"), u256.Zero(), u256.Zero(), false, "", "100000"},
+	}
+
+	for _, tc := range tests {
+		if tc.shouldPanic {
+			uassert.PanicsWithMessage(t, tc.panicMsg, func() { positionUpdate(position, tc.liquidityDelta, tc.feeGrowthInside0X128, tc.feeGrowthInside1X128) })
+		} else {
+			newPos := positionUpdate(position, tc.liquidityDelta, tc.feeGrowthInside0X128, tc.feeGrowthInside1X128)
+			uassert.Equal(t, newPos.liquidity.ToString(), tc.expectedLiquidity)
```

Comment:
positionUpdate plays two important roles.
1. update liquidity for the position, if there is a liquidity delta
2. calculate and update the accumulated fee for the liquidity provided

There is a check for changing liquidity: 0 -> fluctuating liquidity check.
Non-zero liquidity -> changing liquidity check is also needed to identify any problems in the calculation.
We also need to make sure that the accumulated fee is updated.
Both values are very important in a swap, so it's important to make sure.

Reply:
a0ce7d1e80554e281b20bbe5996feadaab954f5f

Reply:
@r3v4s 
not enough

positionUpdate set tokensOwed0 and tokensOwed1
we need to check tokensOwed update

Reply:
66253fc67240b4da9e4e9464351b2fa9cd091a88